examples/list.bend

```diff
@@ -0,0 +1,169 @@
+##############################
+# Author: Ematth, 2024
+##############################
+### Singly-Linked List Type Definition: ###
+
+# type List:
+#   Cons { head, ~tail }
+#   Nil
+
+###########################################
+
+# List clear:
+# List l -> list l
+# clears all elements from list l.
+# (this might be completely useless...)
+List/clear = @l
+  []
+
+# List concat:
+# List l -> List l
+# combines two lists (l1, l2) from left to right.
+List/concat = @l1 @l2
+  match l1 {
+    List/Cons: (List/Cons l1.head (List/concat l1.tail l2))
+    List/Nil: l2
+  }
+
+# List add_front:
+# List l -> List l
+# adds a non-List element e to the front of list l.
+List/add_front = @l @e
+  match l {
+    List/Cons: (List/Cons e l)
+    List/Nil: (List/Cons e List/Nil)
+  }
+
+# List append (add_back):
+# List l -> List l
+# adds a non-list element e to the back of list l.
+List/append = @l @e
+  (List/concat l (List/Cons e List/Nil))
+
+# List reverse:
+# List l -> List l
+# reverses the order of elements in list l.
+List/reverse/aux = @acc @l
+  match l {
+    List/Nil: acc
+    List/Cons: (List/reverse/aux (List/Cons l.head acc) l.tail)
+  }
+List/reverse = @l
+  (List/reverse/aux [] l)
+
+# List length:
+# List l -> uint
+# returns the number of elements in list l.
+List/len/aux = @acc @l
+  match l {
+    List/Nil: acc
+    List/Cons: (List/len/aux (+ acc 1) l.tail)
+  }
+List/len = @l
+  (List/len/aux 0 l)
+
+# List count:
+# List l -> Some s -> uint
+# returns the number of instances of element s in list l.
+List/count/aux = @acc @l @s
+  match l {
+    List/Nil: acc
+    List/Cons: use acc = switch (== l.head s) {
+      0: acc;
+      _: (+ acc 1);
+    }
+    (List/count/aux acc l.tail s)
+  }
+List/count = @l @s
+  (List/count/aux 0 l s)
+
+# List index:
+# List l -> Some s
+# returns the value of a specific list index i, or * if the index doesn't exist.
+List/index = @l @i
+  match l {
+    List/Cons: 
+      switch i {
+        0: l.head
+        _: (List/index l.tail (- i 1))
+      }
+    List/Nil: *
+  }
+
+# List pop_front: 
+# List l -> (List l, Some s)
+# removes the first item from the front of list l, or [] if the list is empty.
+# Both the popped element and the new list are returned.
+List/pop_front = @l
+  match l {
+    List/Cons: (l.head, l.tail) 
+    List/Nil: []
+  }
+
+# List pop_back:
+# List l -> List l
+# removes the first item from the back of list l.
+List/pop_back/aux = @acc @l
+  match l {
+    List/Nil: List/Nil
+    List/Cons: 
+      use x = switch acc {
+        0: *
+        _: l.head
+      }
+      use y = switch acc {
+        0: List/Nil
+        1: List/Nil
+        _: l.tail
+      }
+      (List/Cons x (List/pop_back/aux (- acc 1) y))
+  }
+List/pop_back = @l
+  switch (- (List/len l) 1) {
```

Comment:
This is a great example of a case where List/len should be linear, otherwise this becomes quadratic in complexity (for each element you're duplicating all the rest of the list just to check its length)